contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,1153 @@
+        );
+
+        if (owedTokenToWithdraw == 0) {
+            return 0;
+        }
+
+        // check that there is enough token to give back
+        require(
+            owedTokenToWithdraw <= availableForBucket[bucket],
+            "BucketLender#withdrawOwedToken: There must be enough available owedToken"
+        );
+
+        // update amounts
+        updateAvailable(bucket, owedTokenToWithdraw, false);
+
+        return owedTokenToWithdraw;
+    }
+
+    /**
+     * Helper function to withdraw heldToken from this contract.
+     *
+     * @param  bucket        The bucket number to withdraw from
+     * @param  userWeight    The amount of weight the user is using to withdraw
+     * @param  bucketWeight  The total weight of the bucket
+     * @param  maxHeldToken  The total amount of heldToken available to withdraw
+     * @return               The amount of heldToken being withdrawn
+     */
+    function withdrawHeldToken(
+        uint256 bucket,
+        uint256 userWeight,
+        uint256 bucketWeight,
+        uint256 maxHeldToken
+    )
+        private
+        returns (uint256)
+    {
+        if (maxHeldToken == 0) {
+            return 0;
+        }
+
+        // user's principal for the bucket
+        uint256 principalForBucketForAccount = MathHelpers.getPartialAmount(
+            userWeight,
+            bucketWeight,
+            principalForBucket[bucket]
+        );
+
+        if (principalForBucketForAccount == 0) {
+            return 0;
+        }
+
+        uint256 heldTokenToWithdraw = MathHelpers.getPartialAmount(
+            principalForBucketForAccount,
+            principalTotal,
+            maxHeldToken
+        );
+
+        updatePrincipal(bucket, principalForBucketForAccount, false);
+
+        return heldTokenToWithdraw;
+    }
+
+    // ============ Setter Functions ============
+
+    /**
+     * Changes the critical bucket variable
+     *
+     * @param  bucket  The value to set criticalBucket to
+     */
+    function setCriticalBucket(
+        uint256 bucket
+    )
+        private
+    {
+        // don't spend the gas to sstore unless we need to change the value
+        if (criticalBucket != bucket) {
+            criticalBucket = bucket;
+        }
+    }
+
+    /**
+     * Changes the available owedToken amount. This changes both the variable to track the total
+     * amount as well as the variable to track a particular bucket.
+     *
+     * @param  bucket    The bucket number
+     * @param  amount    The amount to change the available amount by
+     * @param  increase  True if positive change, false if negative change
+     */
+    function updateAvailable(
+        uint256 bucket,
+        uint256 amount,
+        bool increase
+    )
+        private
+    {
+        if (increase) {
+            availableTotal = availableTotal.add(amount);
+            availableForBucket[bucket] = availableForBucket[bucket].add(amount);
```

Comment:
I don't think an assert here is that useful. Since we are adding/subtracting the same amount from both, if the assert were to fail, that means that the values were already messed up. So an assert wouldn't catch any mess-ups